### PR TITLE
Issues/23978

### PR DESCRIPTION
**Pull request by aseren**
Added triggering of the notification timed-out event when custom ad notifications are disabled.

Resolves https://github.com/brave/brave-browser/issues/23978

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

For Desktop:
- Trigger Native Brave ad notification  on one of the desktop platforms (Windows, MacOS, Linux)
- Wait until the ad notification is closed by timeout
- Check that the time-out event was triggered (`Timed out notification ad with placement id ` message in rewards-log)

For Android:
- Check that Brave ad notification is shown on Android

### CODE REVIEW

File: browser/brave_ads/notification_helper/notification_helper.h [hunk 1]
```diff
@@ -1,32 +1,52 @@
...
+
+ private:
+  friend struct base::DefaultSingletonTraits<NotificationHelper>;
 
   NotificationHelper();
+  ~NotificationHelper();
+
+  void OnSystemNotificationPlatformBridgeReady(bool success);
+
+  bool system_notifications_supported_ = true;
```
Comment (tmancey) [inline browser/brave_ads/notification_helper/notification_helper.h:45]: nit: `does_supported_native_notifications_`

File: browser/brave_ads/notification_helper/notification_helper.h [hunk 2]
```diff
@@ -1,32 +1,52 @@
...
 
-  virtual bool ShowOnboardingNotification();
+  bool CanShowNativeNotifications();
+  bool CanShowNativeNotificationsWhileBrowserIsBackgrounded() const;
 
- protected:
-  friend class NotificationHelperHolder;
+  bool ShowOnboardingNotification();
+
+  bool SystemNotificationsSupported() const;
```
Comment (tmancey) [inline browser/brave_ads/notification_helper/notification_helper.h:35]: Could this not be part of `CanShowNativeNotifications`? As CanShow should only return true if native/system notifications are supported

**Review by tmancey — APPROVED** [review]
LGTM++ (see comments)

File: browser/brave_ads/notification_helper/notification_helper.h (hunk 2 above)
Comment (aseren) [reply browser/brave_ads/notification_helper/notification_helper.h:35]: Renamed `CanShowNativeNotifications` to `CanShowNotifications` as it is true for both system and chrome message center notifications.
Renamed `CanShowNativeNotificationsWhileBrowserIsBackgrounded` to `CanShowSystemNotificationsWhileBrowserIsBackgrounded`.

**Comment by btlechowski** [comment]
Verification passed on

Brave | 1.44.18 Chromium: 104.0.5112.81 (Official Build) nightly (64-bit)
-- | --
Revision | 5b7b76419d50f583022568b6764b630f6ddc9208-refs/branch-heads/5112@{#1309}
OS | Ubuntu 18.04 LTS

Verified timeout was scheduled for notification ad:
```
[2553:2553:0803/200944.217458:VERBOSE1:ads_service_impl.cc(929)] Timeout notification ad with placement id 4aba3a66-19ed-4505-a7e0-a89a1d6e0e93 in 120 seconds
```

Verified timeout was performed and ad disappeared.
```
[2553:2553:0803/201144.228039:VERBOSE3:notification_ad_event_timed_out.cc(21)] Timed out notification ad with placement id 4aba3a66-19ed-4505-a7e0-a89a1d6e0e93 and creative instance id c80153f6-cf22-41c0-8c2c-24c1f9a23c96
```

**Pull request merged**